src/Npgsql/NpgsqlSchema.cs

```diff
@@ -438,6 +439,291 @@ public static DataTable GetReservedWords()
             return table;
         }
 
+        internal static DataTable GetDataTypes()
+        {
+            var table = new DataTable("DataTypes");
+            table.Locale = CultureInfo.InvariantCulture;
+            table.Columns.Add("TypeName", typeof(string));
+            table.Columns.Add("ProviderDbType", typeof(string));
+            table.Columns.Add("ColumnSize", typeof(long));
+            table.Columns.Add("CreateFormat", typeof(string));
+            table.Columns.Add("CreateParameters", typeof(string));
+            table.Columns.Add("DataType", typeof(string));
+            table.Columns.Add("IsAutoincrementable", typeof(bool));
+            table.Columns.Add("IsBestMatch", typeof(bool));
+            table.Columns.Add("IsCaseSensitive", typeof(bool));
+            table.Columns.Add("IsFixedLength", typeof(bool));
+            table.Columns.Add("IsFixedPrecisionScale", typeof(bool));
+            table.Columns.Add("IsLong", typeof(bool));
+            table.Columns.Add("IsNullable", typeof(bool));
+            table.Columns.Add("IsSearchable", typeof(bool));
+            table.Columns.Add("IsSearchableWithLike", typeof(bool));
+            table.Columns.Add("IsUnsigned", typeof(bool));
+            table.Columns.Add("MaximumScale", typeof(short));
+            table.Columns.Add("MinimumScale", typeof(short));
+            table.Columns.Add("IsConcurrencyType", typeof(bool));
+            table.Columns.Add("IsLiteralSupported", typeof(bool));
+            table.Columns.Add("LiteralPrefix", typeof(string));
+            table.Columns.Add("LiteralSuffix", typeof(string));
+            table.Columns.Add("NativeDataType", typeof(string));
+
+
+            var obj = new object[40, 23] {
+            {"CHAR", (int)NpgsqlDbType.Char, DBNull.Value, "CHAR({0})",
+                    "length", "System.String",
+                    false, false, true, true, false,
+                    false, true, true, true, DBNull.Value,
+                    DBNull.Value, DBNull.Value,
+                    false, true, "'", "'", "DBTYPE_STR"},
```

Comment:
It would be much better if you construct an array of `ValueTuple`s and copy them using the generic method below:

```csharp
foreach (var type in Types)
{
    table.Rows.Add(ConstructRow(type));

    DataRow ConstructRow<T>(T tuple)
        where T : struct, ITuple
    {
        var row = table.NewRow();
        for (var i = 0; i < type.Length; i++)
            row[j] = tuple[j] ?? DbNull.Value;
    }
}
```

Reply:
@YohDeadfall isn't it simpler to simply construct the object array statically, ensuring it's allocated only once? A tuple or an object array is frankly almost the same thing in this context...

Reply:
Only if he box `bool` values once before building the array and use them later to store in it:

```csharp
static readonly object TrueBox = true;
static readonly object FalseBox = false;
```

The same for lengths.

Reply:
Are you sure the runtime doesn't optimize the use of boxed bool literals to not allocate an instance for each? For example, when an async method returning `Task<bool>` returns synchronously it does not allocate a Task, but has one global static true Task and another false Task which are returned.

But even if boxing occurs each time, it's a one-time allocation, so there's very little impact on anything... Also, unless I'm mistaken, the DataRow contains objects, so in any case the bools will get boxed. In fact, if the method is called several times, then the tuple's bool will get boxed each time, but an the array boxes only once at initialization and we copy the references...

In any case this conversation is more for fun, the impact is negligible :)

Reply:
Unfortunately, no. Probably, in UWP methods of the `PropertyValue` class could produce singletons for boxed values, but I'm not sure.

```csharp
Console.WriteLine(ReferenceEquals(true, true)
    ? "Yay! It produces the same boxed object."
    : "The compiler isn't smart enough.");
```

By the way, you can find that WPF uses `BooleanBoxes` internally to reuse boxed values.

Reply:
I'm sure I saw an issue discussing this but no matter.

Thinking about this again, assuming we hard-code all of these details, why not simply create a DataTable every time with all the values in code and return it - the current approach of copying from some other static structure (whether array of objects or tuples) doesn't seem to add much.

Let's see if @warcha intends to fix the comments, if not I'll sit down and look at doing this in a nicer way, including getting at least some info out of the loaded types/mappings.